shared/entity/type.go

```diff
@@ -338,3 +340,89 @@ entityTypeLoop:
+// empty string is return and the error is logged with as much context as possible, including the file and line number
+// of the caller.
+func (t Type) urlMust(projectName string, location string, pathArguments ...string) *api.URL {
+	ref, err := t.URL(projectName, location, pathArguments...)
+	if err != nil {
+		logCtx := logger.Ctx{"entity_type": t, "project_name": projectName, "location": location, "path_aguments": pathArguments}
+
+		// Get the second caller (we expect the first caller to be internal to this package since this method is not exported).
+		_, file, line, ok := runtime.Caller(2)
+		if ok {
+			logCtx["caller"] = fmt.Sprintf("%s#%d", file, line)
+		}
+
+		logger.Error("Failed to create entity URL", logCtx)
+		return api.NewURL()
+	}
+
+	return ref
+}
+
+// ProjectURL returns an *api.URL to a Project.
+func ProjectURL(projectName string) *api.URL {
+	return TypeProject.urlMust("", "", projectName)
+}
+
+// InstanceURL returns an *api.URL to an instance.
+func InstanceURL(projectName string, instanceName string) *api.URL {
+	return TypeInstance.urlMust(projectName, "", instanceName)
+}
+
+// ServerURL returns an *api.URL to the server.
+func ServerURL() *api.URL {
+	return TypeServer.urlMust("", "")
+}
+
+// CertificateURL returns an *api.URL to a certificate.
+func CertificateURL(fingerprint string) *api.URL {
+	return TypeCertificate.urlMust("", "", fingerprint)
+}
+
+// ImageURL returns an *api.URL to an image.
+func ImageURL(projectName string, imageName string) *api.URL {
+	return TypeImage.urlMust(projectName, "", imageName)
+}
+
+// ImageAliasURL returns an *api.URL to an image alias.
+func ImageAliasURL(projectName string, imageAliasName string) *api.URL {
+	return TypeImageAlias.urlMust(projectName, "", imageAliasName)
+}
+
+// ProfileURL returns an *api.URL to a profile.
+func ProfileURL(projectName string, profileName string) *api.URL {
+	return TypeProfile.urlMust(projectName, "", profileName)
+}
+
+// NetworkURL returns an *api.URL to a network.
+func NetworkURL(projectName string, networkName string) *api.URL {
+	return TypeNetwork.urlMust(projectName, "", networkName)
+}
+
+// NetworkACLURL returns an *api.URL to a network ACL.
+func NetworkACLURL(projectName string, networkACLName string) *api.URL {
+	return TypeNetworkACL.urlMust(projectName, "", networkACLName)
+}
+
+// NetworkZoneURL returns an *api.URL to a network zone.
+func NetworkZoneURL(projectName string, networkZoneName string) *api.URL {
+	return TypeNetworkZone.urlMust(projectName, "", networkZoneName)
+}
+
+// StoragePoolURL returns an *api.URL to a storage pool.
+func StoragePoolURL(storagePoolName string) *api.URL {
+	return TypeStoragePool.urlMust("", "", storagePoolName)
+}
+
+// StorageVolumeURL returns an *api.URL to a storage volume.
+func StorageVolumeURL(projectName string, location string, storagePoolName string, storageVolumeType string, storageVolumeName string) *api.URL {
```

Comment:
This is starting to duplicate the use of the `shared/api/*` struct's `URL()` function, e.g. 

https://github.com/canonical/lxd/blob/main/shared/api/storage_pool_volume.go#L144
https://github.com/canonical/lxd/blob/main/shared/api/storage_pool_bucket.go#L75

Its an approach not wholly supported by all api struct entity types yet, but there seems to be some sense (to me anyway) in having the the public API struct entity types be able to know how to generate their own URLs.

As this likely is going to be not in shared now, it doesn't have to change what we are doing here at this time.

But given your plan to remove duplication in https://github.com/canonical/lxd/issues/12868 i wanted to point out there is another level of duplication starting to occur here too.

Reply:
I didn't see these and I agree. I'll add a link to this comment to that issue.